source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc

```diff
@@ -27,19 +35,28 @@ Http::FilterHeadersStatus GcpAuthnFilter::decodeHeaders(Http::RequestHeaderMap&
   Envoy::Upstream::ThreadLocalCluster* cluster =
       context_.clusterManager().getThreadLocalCluster(route->routeEntry()->clusterName());
 
-  std::string audience_str;
   if (cluster != nullptr) {
     // The `audience` is passed to filter through cluster metadata.
     auto filter_metadata = cluster->info()->metadata().typed_filter_metadata();
     const auto filter_it = filter_metadata.find(std::string(FilterName));
     if (filter_it != filter_metadata.end()) {
       envoy::extensions::filters::http::gcp_authn::v3::Audience audience;
       MessageUtil::unpackTo(filter_it->second, audience);
-      audience_str = audience.url();
+      audience_str_ = audience.url();
     }
   }
 
-  if (!audience_str.empty()) {
+  if (!audience_str_.empty()) {
+    if (jwt_token_cache_ != nullptr) {
+      auto token = jwt_token_cache_->lookUp(audience_str_);
+      if (token != nullptr) {
+        // If token is found in the cache, we add the token string to the request directly and
+        // continue the filter chain iteration.
+        addTokenToRequest(hdrs, token->jwt_);
+        return FilterHeadersStatus::Continue;
+      }
+    }
+
```

Comment:
You may want to add a mechanism to track if request for the specific audience has already been made. Otherwise if metadata service is slow Envoy will keep piling up gRPC requests for the same audience.